File: Queue/Queue.java
package Queue;

public class Queue<T> {

    public static class Node<T> {
        private T data;
        private Node<T> next;

        public Node(T data){
            this.data = data;
            this.next = null;
        }

        public T returnData(){
            return this.data;
        }

    }

    private Node<T> top;
    private Node<T> bottom;
    public Queue(T data){
        this.top = new Node<T>(data);
        this.bottom = top;
    }

    public Queue(){
        this.top = null;
        this.bottom = null;
    }

    public T peek(){
        if(top == null){
            return null;
        }
        return this.top.data;
    }

    public void add(T data){
        Node<T> node = new Node<T>(data);
        if(this.bottom != null){
            this.bottom.next = node;
        }
        this.bottom = node;
        if(this.top == null){
            this.top = this.bottom;
        }
        return;
    }

    public void add(Node<T> node){
        if(this.bottom != null){
            this.bottom.next = node;
        }
        this.bottom = node;
        if(this.top == null){
            this.top = this.bottom;
        }
        return;
    }
    
    public T remove(){
        if(top == null){
            return null;
        }
        T value = this.top.data;
        this.top = this.top.next;
        if(this.top == null){
            this.bottom = null;
        }
        return value;
    }

    public Node<T> removeNode(){
        if(top == null){
            return null;
        }
        Node<T> value = this.top;
        this.top = this.top.next;
        if(this.top == null){
            this.bottom = null;
        }
        return value;
    }

    public boolean isEmpty(){
        return this.top == null;
    }

    public void print(){
        Node<T> n = this.top;
        while(n != null){
            System.out.print(n.data+" ");
            n = n.next;
        }
        System.out.println();
        return;
    }

}

File: LeetCodePython/BasicStructures/Queue.py
#BasicQueue
class Queue:
    def __init__(self, node = None):
        if node == None:
            self.head = None
            self.tail = None
            return
        elif not isinstance (node, Node):
            node = Node(node)
        self.head = node
        self.tail = node

    def __iter__(self):
        n = self.head
        while n:
            yield repr(n)
            n = n.next

    def add(self, node):
        if not isinstance(node, Node):
            node = Node(node)
        
        if self.head == None:
            self.head = node
            self.tail = node
            return
        
        self.tail.next = node
        self.tail = node
    
    def pop(self):
        if self.head == None:
            return None
        
        if self.head == self.tail:
            self.head = None
            self.tail = None
            return None
        value = repr(self.head)
        self.head = self.head.next
        return value

    def peek(self):
        return repr(self.head)
    
    def to_string(self):
        return ''.join(iter(self))

    def is_empty(self):
        return self.head == None


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
    def __repr__(self):
        return repr(self.data)